satpy/readers/fci_l1c_fdhsi.py

```diff
@@ -61,49 +80,75 @@ def end_time(self):
 
     def get_dataset(self, key, info=None):
         """Load a dataset."""
-        if key in self.cache:
-            return self.cache[key]
 
-        logger.debug('Reading {}'.format(key.name))
+        logger.debug('Reading {} from {}'.format(key.name, self.filename))
         # Get the dataset
         # Get metadata for given dataset
-        variable = self.nc['/data/{}/measured/effective_radiance'
-                           .format(key.name)]
-        # Convert to xarray
-        radiances = xr.DataArray(np.asarray(variable, np.float32), dims=['y', 'x'])
-        radiances.attrs['scale_factor'] = variable.attrs['scale_factor']
-        radiances.attrs['offset'] = variable.attrs.get('add_offset', 0)
-        radiances.attrs['FillValue'] = variable.attrs['_FillValue']
-        # Set invalid values to NaN
-        radiances.values[radiances == radiances.attrs['FillValue']] = np.nan
-        # Apply scale factor and offset
-        radiances = radiances * (radiances.attrs['scale_factor'] * 1.0) + radiances.attrs['offset']
-
-        # TODO: Calibration is disabled, waiting for calibration parameters from EUMETSAT
-        res = self.calibrate(radiances, key)
-
-        self.cache[key] = res
-        self.nlines, self.ncols = res.shape
+        measured, root = self.get_channel_dataset(key.name)
+        radlab = measured + "/effective_radiance"
+        data = self[radlab]
+
+        attrs = data.attrs.copy()
+        info = info.copy()
+        fv = attrs.pop(
+                "FillValue",
+                default_fillvals.get(data.dtype.str[1:], np.nan))
+        vr = attrs.pop("valid_range", [-np.inf, np.inf])
+        if key.calibration == "counts":
+            attrs["_FillValue"] = fv
+            nfv = fv
+        else:
+            nfv = np.nan
+        data = data.where(data > vr[0], nfv)
+        data = data.where(data < vr[1], nfv)
+        if key.calibration == "counts":
+            # from package description, this just means not applying add_offset
+            # and scale_factor
+            attrs.pop("scale_factor")
+            attrs.pop("add_offset")
+            data.attrs["units"] = "1"
+            res = data
+        else:
+            data = (data * attrs.pop("scale_factor", 1) +
+                    attrs.pop("add_offset", 0))
+
+            if key.calibration in ("brightness_temperature", "reflectance"):
+                res = self.calibrate(data, key, measured, root)
+            else:
+                res = data
+                data.attrs["units"] = attrs["units"]
+        # pre-calibration units no longer apply
+        info.pop("units")
+        attrs.pop("units")
 
+        self.nlines, self.ncols = res.shape
+        res.attrs.update(key.to_dict())
+        res.attrs.update(info)
+        res.attrs.update(attrs)
```

Comment:
You may be able to avoid some of the `pops` (ex units) if you rearrange when and how the attributes are updated. This isn't a required change, but if you see what I mean and think the code would look prettier then maybe it is worth it.